components/ai_chat/core/browser/text_embedder.cc

```diff
@@ -0,0 +1,308 @@
+/* Copyright (c) 2024 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/ai_chat/core/browser/text_embedder.h"
+
+#include <algorithm>
+
+#include "base/check_is_test.h"
+#include "base/files/file_util.h"
+#include "base/hash/hash.h"
+#include "base/logging.h"
+#include "base/memory/ptr_util.h"
+#include "base/metrics/histogram_functions.h"
+#include "base/strings/strcat.h"
+#include "base/strings/string_split.h"
+#include "base/task/bind_post_task.h"
+#include "base/timer/elapsed_timer.h"
+#include "third_party/tflite/src/tensorflow/lite/core/api/op_resolver.h"
+#include "third_party/tflite_support/src/tensorflow_lite_support/cc/task/text/utils/text_op_resolver.h"
+
+using TFLiteTextEmbedder = tflite::task::text::TextEmbedder;
+using TFLiteTextEmbedderOptions = tflite::task::text::TextEmbedderOptions;
+using tflite::task::text::CreateTextOpResolver;
+
+namespace ai_chat {
+
+size_t TextEmbedder::g_segment_size_ceiling_(300);
+size_t TextEmbedder::g_segment_size_floor_(50);
+
+// static
+std::unique_ptr<TextEmbedder, base::OnTaskRunnerDeleter> TextEmbedder::Create(
+    const base::FilePath& model_path,
+    scoped_refptr<base::SequencedTaskRunner> embedder_task_runner) {
+  if (model_path.empty()) {
+    return std::unique_ptr<TextEmbedder, base::OnTaskRunnerDeleter>(
+        nullptr, base::OnTaskRunnerDeleter(embedder_task_runner));
+  }
+  auto embedder = std::unique_ptr<TextEmbedder, base::OnTaskRunnerDeleter>(
+      new TextEmbedder(model_path, embedder_task_runner),
+      base::OnTaskRunnerDeleter(embedder_task_runner));
+  return embedder;
+}
+
+TextEmbedder::TextEmbedder(
+    const base::FilePath& model_path,
+    scoped_refptr<base::SequencedTaskRunner> embedder_task_runner)
+    : model_path_(model_path),
+      owner_task_runner_(base::SequencedTaskRunner::GetCurrentDefault()),
+      embedder_task_runner_(embedder_task_runner) {
+  DCHECK(owner_task_runner_->RunsTasksInCurrentSequence());
+}
+
+TextEmbedder::~TextEmbedder() = default;
+
+bool TextEmbedder::IsInitialized() const {
+  DCHECK(owner_task_runner_->RunsTasksInCurrentSequence());
+  base::AutoLock auto_lock(lock_);
+  return tflite_text_embedder_ != nullptr;
+}
+
+void TextEmbedder::Initialize(base::OnceCallback<void(bool)> callback) {
+  DCHECK(owner_task_runner_->RunsTasksInCurrentSequence());
+
+  embedder_task_runner_->PostTask(
+      FROM_HERE,
+      base::BindOnce(&TextEmbedder::InitializeEmbedder,
+                     weak_ptr_factory_.GetWeakPtr(),
```

Comment:
also using WeakPtr here is not useful unless you destroy/invalidate `weak_ptr_factory_` on `embedder_task_runner_` and could DCHECK if a task is running on `embedder_task_runner_` when `weak_ptr_factory_` gets destroyed on the owner sequence

Reply:
oh wait, nvm, I forgot this class is destroyed on the embedder sequence so the `weak_ptr_factory_` is as well.